### PR TITLE
Replace unittest assert methods by assert statement

### PR DESCRIPTION
**Pull request by snejus**
... continuing the test suite migration to `pytest` (#5361), this PR replaces `unittest` method assertions with `assert` statement.

See the table below for a list of the replaced methods and their counts. This should help clarify the size of the pull request :laughing:.

To keep things simple for the review, I have replaced each method in its own commit. This
way, things should be easy to follow if you step through the commits one by one. :slightly_smiling_face:

method                   | count
---                      | ---
**`assertEqual`**        | 1747
**`assertIn`**           | 200
**`assertTrue`**         | 149
**`assertIsNone`**       | 87
**`assertFalse`**        | 85
**`assertRaises`**       | 69
**`assertNotIn`**        | 64
**`assertNotEqual`**     | 39
**`assertIsNotNone`**    | 35
**`assertIsInstance`**   | 35
**`assertLessEqual`**    | 23
**`assertGreater`**      | 15
**`assertLess`**         | 14
**`assertGreaterEqual`** | 14
**`assertAlmostEqual`**  | 9
**`assertRaisesRegex`**  | 4
**`assertCountEqual`**   | 2
**`assertListEqual`**    | 1

:exclamation: Note that this *only* replaces methods provided by `unittest` by default.
**Custom** assertion method replacements (like `assertExists` will be addressed in a separate PR).

### CODE REVIEW

**Comment by Serene-Arc** [comment]
For a PR like this, if it's done with like a grep and the tests still pass, I don't think it requires more review than a sanity check. Is that what you did, or did you go through by hand?

**Comment by snejus** [comment]
> For a PR like this, if it's done with like a grep and the tests still pass, I don't think it requires more review than a sanity check. Is that what you did, or did you go through by hand?

Most of it was done using a several `sed` search and replace patterns (see below). Then I later discovered a tool [unittest2pytest](https://pypi.org/project/unittest2pytest/) which I used to double-check my replacements.

I did review the diff manually though since it failed to replace some of the more complicated assertions where I saw the tests fail.


```sh
oneline_python() {
  # args: <python-file> ...
  # info: _unformat given .py files: remove newlines from each statement
  sed -zr '
    s/(^|\n) *# [^\n]*//g
    s/  # [^\n]*//g
    s/,?\n\s*([]}).]+)/\1/g
    s/\n\s+(\b(and|or)\b|==)/ \1/g
    s/,\s*\n\s+/, /g
    s/"\s*\n\s+(%|f['\''"])/" \1/g
    s/([[{(])\s*\n\s+/\1/g
    s/(["'\''])\n\s*(["'\''+])/\1 \2/g
    s/\n\s*( \+)/\1/g
    s/(\[[^]\n]+)\n\s*( for)/\1\2/g
  ' $@ |
    tr '\000' '\n'
}

gdfd() {
  # args: <git-args> <pattern>
  # info: show hunks that match the _<pattern>
  local argidx
  argidx=${@[(rI)[^-]*]}

  git diff -U1 ${@[1, $(( argidx - 1 ))]} |
    grepdiff --output-matching=hunk --only-match=mod ${@[$argidx,-1]} | {
      if [[ -t 1 ]]; then
        diff-so-fancy
      else
        cat
      fi
    }
}

unittest2pytest() {
  # args: <python-file> ...
  # info: convert _unittest assertions to _assert statements
  local file
  local contents
  # black -C --line-length=100000 $@
  for file in $@; do
    oneline=$(oneline_python $file)
    print -R $oneline > $file
    gdfd -U0 $file '  # ' | git apply -R --unidiff-zero --allow-empty
    sed -r '
      /def |assert(Raises|Is(Ins|File|Dir)|Counts|(Not)?(Exist|InResult)|EqualTimes|AlbumImport|Perms|Contains|(No)?File|Item|Excludes|Negation|LyricsCon)/!{
        /[a-z]+.assert[A-Z]/{
          s/\(/ /
          s/\)(  #|$)/\1/
          s/msg=//
          s/[a-z._]*assertTrue (([^()]|\([^)]*\))+)/assert \1/
          s/[a-z._]*assertFalse (([^()]|\([^)]*\))+)/assert not \1/
          s/[a-z._]*assertIsNone (.*)(, |$)/assert \1 is None/
          s/[a-z._]*assertIsNotNone (.*)(, |$)/assert \1 is not None/
          s/[a-z._]*assert(List)?Equal (([^,]|\([^)]*\)|\{[^}]*\}|\[[^\]]*\])+), (#[^@]+@ )?(([^,]|\([^)]*\)|\[[^]]*\])+)/assert \2 == \5/
          s/[a-z._]*assertNotIn ([^,]+), (.*)$/assert \1 not in \2/
          s/[a-z._]*assertIn ([^,]+), (.*)$/assert \1 in \2/
          s/[a-z._]*assertLess ([^,]+), (([^()]|\([^)]*\))+)$/assert \1 < \2/
          s/[a-z._]*assertLessEqual ([^,]+), (([^()]|\([^)]*\))+)$/assert \1 <= \2/
          s/[a-z._]*assertGreaterEqual ([^,]+), (.*)$/assert \1 >= \2/
          s/[a-z._]*assertGreater ([^,]+), (.*)$/assert \1 > \2/
          s/[a-z._]*assertNotEqual ([^,]+), (.*)$/assert \1 != \2/
          # s/[a-z._]*assertExists ([^)]+)/assert Path(\1).exists()/
          # s/[a-z._]*assertNotExists ([^)]+)/assert not Path(\1).exists()/
          /[a-z._]*assertAlmostEqual ([^,]+), ([^,)]+(, (places|delta)=[0-9]+)?)/{
            s//assert \1 == pytest.approx(\2)/
            s/places=/rel=1e-/
            s/delta=/abs=/
          }
          # s/[a-z._]*assertEqualTimes ([^,]+), ([^,]+)/assert \1 == pytest.approx(\2, rel=1e-4)/
          # s/[a-z._]*assertInResult ([^,]+), (.*)/assert \1.id in {i.id for i in \2}/
          # s/[a-z._]*assertNotInResult ([^,]+), (.*)/assert \1.id not in {i.id for i in \2}/
        }
      }
      s/[a-z._]*assertRaisesRegex(\([^,]+, )/pytest.raises\1match=/
      /[a-z._]*assertRaises/{
        s//pytest.raises/
        s/^(( +)([a-z][^,]+)), ([^,]+)(, ([^)]+))?\)$/\2with \3):\n\2    \4(\6)/
      }
      s/[a-z._]*assertIsInstance/assert isinstance/
    ' -i $file
  done
  poe format $@
}
```

**Comment by Serene-Arc** [comment]
In that case I'll give all of it a quick lookover.

**Review by bal-e — COMMENTED** [review]
Great job, @snejus!

**Review by Serene-Arc — APPROVED** [review]

File: test/plugins/test_advancedrewrite.py [hunk 1]
```diff
@@ -84,19 +86,19 @@ def test_advanced_rewrite_example_with_multi_valued_field(self):
...
 
-            self.assertEqual(item.artists, ["유빈", "미미"])
+            assert item.artists == ["유빈", "미미"]
 
     def test_fail_when_replacements_empty(self):
-        with self.assertRaises(
+        with pytest.raises(
             UserError,
-            msg="Advanced rewrites must have at least one replacement",
+            match="Advanced rewrites must have at least one replacement",
```
[the remark this answers is not in the document]
Comment (Serene-Arc) [reply test/plugins/test_advancedrewrite.py:94]: Yeah, this is the only thing that I found. Everything else is equivalent to the previous tests as far as I can see.

[the remark this answers is not in the document]
Comment (snejus) [reply test/plugins/test_advancedrewrite.py:94]: Do you reckon we should be asserting _the entire_ exception message?

I tend to assert only the relevant part in the exception message, for example

```python
from contextlib import ExitStack as does_not_raise

import pytest
from django.core.management import call_command
from django.core.management.base import CommandError


def error(message_pattern: str):
    return pytest.raises(CommandError, match=message_pattern)


@pytest.mark.parametrize(
    "ids, expected_error",
    (
        ([], error(r"required: ids"), 0),
        ([0, 1], error(r".*do not exist"), 0),
        ([1, 9], error(r"already have request"), 0),
        ([1, 2], does_not_raise(), 2),
    ),
)
def test_create_leave_requests_for_leave(ids, expected_error):
    with expected_error:
        call_command("my_command", *map(str, ids))
```

this allows the rest of the wording in the message to change. For example, if someone corrects the spelling in the rest of the message, tests won't fail!

File: test/plugins/test_beatport.py [hunk 2]
```diff
@@ -575,22 +569,25 @@ def test_track_url_applied(self):
...
         for track, test_track, id in zip(self.tracks, self.test_tracks, ids):
-            self.assertEqual(
-                track.url,
-                "https://beatport.com/track/" + test_track.url + "/" + str(id),
+            assert (
+                track.url
+                == "https://beatport.com/track/"
+                + test_track.url
+                + "/"
+                + str(id)
```
[the remark this answers is not in the document]
Comment (snejus) [reply test/plugins/test_beatport.py:577]: Makes sense. I'll run [sourcery](https://docs.sourcery.ai/Coding-Assistant/Coding-Assistant-Introduction/) to do the replacements in the code that I changed.

File: test/plugins/test_advancedrewrite.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bal-e) [reply test/plugins/test_advancedrewrite.py:94]: If there's a spelling mistake in the rest of the message, it should be caught by the tests.  If you know what the entire error message should be, and it's at least consistent across platforms, then I think it is best to check for it.

(Localization of user-facing error messages is a different story, one I don't think beets is in any position to think about right now.)

[the remark this answers is not in the document]
Comment (snejus) [reply test/plugins/test_advancedrewrite.py:94]: Localisation would indeed be an interesting issue to tackle, but like you said beets it's in no position to achieve it soon.

The situation I am talking about is something similar to my changes in `autobpm`, for example:

image

If we had two functionally equivalent assertions

```python
assert error == "LibsndfileError: failed to load /a/b/c.mp3"
assert "failed to load" in error.lower()
```

Even if my change kept the relevant information, the first assertion would have started failing, while the second one would have passed.

[the remark this answers is not in the document]
Comment (snejus) [reply test/plugins/test_advancedrewrite.py:94]: I'm merging this in but keeping this thread open for the discussion 😄

**Pull request merged**